### ISSUE TITLE
Data repository management tool required

### ISSUE THREAD

**Issue by jeromekelleher** [ga4gh/server#814]
The current server works well for the example dataset, but this is of limited use beyond development. We need to be able to easily serve arbitrary data from the server. Currently, we require the user to have a clear understanding of how the data repository is organised, and also to create JSON files of a specific format containing MD5 sums and other miscellaneous stuff. This is clearly not acceptable from a usability perspective.

One possible way to resolve this is to provide a data repository management CLI tool, say, ``garepo`` (or ``ga4gh_repo`` if we don't change the naming convention for CLIs as discussed in #813). Example usage:

```
$ garepo init /path/to/example-repo
Empty GA4GH data repository initialised at `/path/to/example-repo`
```
Initialises a new GA4GH data repository at a specified path. This corresponds to creating the directory structure that we are currently using.
```
$ garepo add-dataset /path/to/example-repo dataset1 --description="Example dataset"
Added dataset `dataset1` to /path/to/example-repo
```
Adds a new empty dataset to the repository with the specified name. Any additional metadata can be specified on the command line, as demonstrated here with the ``--description`` argument.

```
$ garepo add-referenceset /path/to/fasta_file --name=GRCh38 /path/to/example-repo/
Added reference set `GRCh38` to GA4GH data repository at `/path/to/example-repo/`
```
Adds a new referenceset based on a given fasta file. This creates the required referenceset directory in the specified repo, creates fasta files, calculates checksums and JSON files, as required. We can add command line options to deal with all the messiness of real data sources.

```
$ garepo add-readgroupset sample_name.bam `/path/to/example-repo/dataset1` --soft-link
Added readGroupSet `sample_name` to dataset `dataset1` in repository `/path/to/example-repo`
```
Adds a new readgroupset to the dataset `dataset1` in  repository `/path/to/example-repo` (note the dataset is provided in the repo URL). Note that we provide the option to soft-link to the BAM file rather than copying it in. If this was a HTTP or FTP URL rather than a local file we'd set up the repository to point to this file using htslib's remote access.

```
$ garepo add-variantset /path/to/vcf/directory  /path/to/example-repo/dataset1 --name=1kg-phase1 --referenceSet=GRCh38 
Added variantSet `1kg-phase1` to dataset `dataset1` in repository `/path/to/example-repo`
```
Add a new variant set to the dataset in the repo.

We'd also want some corresponding commands to remove the various types of object from the repo, as well as a ``garepo list`` or something to write out a description of the repository to stdout.

Thoughts?

**Comment by diekhans** [comment]
merge with https://github.com/ga4gh/server/issues/573

**Comment by macieksmuga** [comment]
The ability to create a reference set based on a single FASTA file would definitely be nice. I got skewered on the weird requirement that each reference FASTA file contain a single reference contig:
```
[Thu Jan 07 06:12:53 2016] [error] [client 50.1.98.156] NotExactlyOneReferenceException: FASTA files must have one and only one reference.  File /mnt/ga4gh/data/ga4gh-count1-data/referenceSets/GRCh38/GCA_000001405.20_GRCh38.p5_genomic.fa.gz has 517 references.
```
Which strikes me as strange, since the single-FASTA distribution of GRCh38 sure is a nice convenient way to portably move around that reference set...

**Comment by diekhans** [comment]
@macieksmuga please create a separate ticket for this enhancement.


This restriction is understandable, as NCBI distributes
the main human genome with one chromosome per fasta:

  ftp://ftp.ncbi.nlm.nih.gov/genomes/Homo_sapiens/CHR_01/

however, when you start throwing in unplaced regions and
altloci, this doesn't really work any more.

samtools provides fasta indexing.  Hopeful part of pysam.

**Comment by macieksmuga** [comment]
Thanks, Mark, I suspected that this restriction may break for more complicated reference structures (it sure does in the graph case). See issue #829

**Comment by dcolligan** [comment]
Considerations mentioned in Friday's call:
- allow an option to specify md5 checksums
- investigate indexing fastas instead of breaking them up
- pull BAMs out of S3 buckets
- option for symlinks
- (server) support for CRAM

**Comment by dcolligan** [comment]
Github erroneously closed this

**Comment by dcolligan** [comment]
Github erroneously closed this, again

**Comment by dcolligan** [comment]
Fixed in #856
